# Exploration of eBay Car Sales Data
Here, we are going to work with a dataset of cars cars from eBay Kleinanzeigen which is a classifieds section of the German eBay website.
The data dictionary is as follows:

* `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
* `name` - Name of the car.
* `seller` - Whether the seller is private or a dealer.
* `offerType` - The type of listing
* `price` - The price on the ad to sell the car.
* `abtest` - Whether the listing is included in an A/B test.
* `vehicleType` - The vehicle Type.
* `yearOfRegistration` - The year in which the car was first registered.
* `gearbox` - The transmission type.
* `powerPS` - The power of the car in PS.
* `model` - The car model name.
* `kilometer` - How many kilometers the car has driven.
* `monthOfRegistration` - The month in which the car was first registered.
* `fuelType` - What type of fuel the car uses.
* `brand` - The brand of the car.
* `notRepairedDamage` - If the car has a damage which is not yet repaired.
* `dateCreated` - The date on which the eBay listing was created.
* `nrOfPictures` - The number of pictures in the ad.
* `postalCode` - The postal code for the location of the vehicle.
* `lastSeenOnline` - When the crawler saw this ad last online.

Our main objective is to clean and analyze the data.

Let's now import the libraries that we are goig to use for this dataset in order to read it into `pandas`.

In [1]:
import numpy as np, pandas as pd

Now let's read the ***auto.csv*** CSV file into pandas

In [2]:
autos = pd.read_csv("autos.csv", encoding = "Latin-1")

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

Let's display the information and list a few data items from our dataset

In [4]:
autos.info()
# autos.shape[0]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [5]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

From the above dataset, I've observed a few things.
* It has a record of ***50000*** rows making it to have ___50000___ entries and 20 columns
* There are some white spaces which we may need to remove from the data in the ***name***, ___vehicle Type___, ***gearbox***, ___model___, ***fuel Type*** and ___brand___ columns
* We may need to edit the column names to be able to use it efficiently
* For the ***price*** column, we will need to remove the dollar signs and commas to be able to work with the price as a number.
* The pricing is in dollars
* We have `integer` and `object` datatypes where most of the columns are of `object` datatypes and only five are of integer datatype (***powerPS***, ___yearOfRegistration___, ***monthOfRegistration***, ___nrOfPictures___ and ***postalCode***)
* Most of the data is stored as strings
* The dates are stored as strings also
* the column names use camelcase and we might need to change that also

## Cleaning the columns
Now we are going to convert the column names from `camelcase` to `snakecase` which is widely prefered in Python.
We are going to change the column names based on our data desctionary to give the names meaningful descriptions

In [6]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [7]:
# columns_copy = autos.columns.copy()
# columns_copy

Now let's change the column names from camelcase to snakecase and assigning back to ***autos.columns***.

In [8]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year ', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']

In [9]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year   \
0  privat    Angebot  $5,000  control          bus                2004   
1  privat    Angebot  $8,500  control    limousine                1997   
2  privat    Angebot  $8,990     test    limousine                2009   
3  privat    Angebot  $4,350  control   kleinwagen                2007   
4  privat    Angebot  $1,350     test        kombi                2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

We changed the column names from camelcase to snakecase and at the same time making them descriptive so that we may be able to work with the dataset in pandas. 

Now, we'll look for columns that have same or almost same values which may be dropped from the dataset because they are not of help to us to be able to provide use ful information for analysis. <br>
We'll also look at the columns that store nimeric data as text, where we'll find which ones are to be cleaned and converted.<br>
We'll use some few nmethods that are very useful in this scenario:
* `DataFrame.describe()` which we will use with (`include='all'`) to get both ___categorical___ and ___numeric___ columns
* `Series.value_counts()` and `Series.head()` to find out if any columns need to be looked at more closely.

Let's use ***autos.describe()*** to look at the descriptive statistics for our columns.

In [10]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-12 16:06:22  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year   gearbox      power_ps  model  \
count         44905        50000.000000    47320  50000.000000  47242   
unique            8                 NaN        2           NaN    245   
top       limousine                 NaN  manuell           NaN   golf   
freq          12859                 NaN    36993           NaN   4024   
mean            NaN         2005.073280      NaN    116.355920    NaN   
std             NaN          105.712813      NaN    209.216627    NaN   
min             NaN         1000.000000      NaN      0.000000    NaN   
25%             NaN         1999.000000      NaN     70.000000    NaN   
50%             NaN         2003.000000      NaN    105.000000    NaN   
75%             NaN         2008.000000      NaN    150.000000    NaN   
max             NaN         9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

From the dataset above:
* Each column has a count of ***50000*** records and colums such as `seller` and `offer_type` have almost similar records.
* The `num_photos` column looks very funny and needs some further looking into.
* `postal_code`, `odometer` and `price` column data are examples of numeric data stored as text that need to be cleaned and combverted.
* The price column has numeric data stored as text with dollar symbols. We will also look into that and convert it to numeric data in order to work with it.

Removing ***non-numeric*** characters from the ___price___ and ___odometer___ columns and cinverting the data to numeric for each column 

In [11]:
# for price column
autos["price"] = (autos['price'].str.replace("$", "").str.replace(",", "").astype(int))
autos["price"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [12]:
# for odometer column
autos["odometer"] = (autos["odometer"].str.replace("km", "").str.replace(",", "").astype(int))
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)
autos["odometer_km"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

In [13]:
# confirming changes have been made for the columns above
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price  ab_test vehicle_type  registration_year   \
0  privat    Angebot   5000  control          bus                2004   
1  privat    Angebot   8500  control    limousine                1997   
2  privat    Angebot   8990     test    limousine                2009   
3  privat    Angebot   4350  control   kleinwagen                2007   
4  privat    Angebot   1350     test        kombi                2003   

     gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere       150000                   3       lpg   
1  automatik       286     7er       150000                   6    benzin   
2    manuell       102    golf        70000                   7    benzin   
3  automatik        71  fortwo        70000                   6    benzin   
4    manuell         0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

Now since we don't need the columns with weird data (`seller`, `offer_type`, 'num_photos'), we will remove them from our dataset since they will not be of much help when analyzing the data.

In [14]:
# starting eith num_photos column
autos["num_photos"].value_counts()

0    50000
Name: num_photos, dtype: int64

As you can see, the column has got only ___zeros___. This is of no help to us during analysis.


The `seller` and `offer_type` columns have almost similar/one value of data, of which is of no help to us also

In [15]:
# let's check for seller now
autos["seller"].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [16]:
# for offer_type now
autos["offer_type"].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

As you can see from the results for `seller` and `offer_type` above, the data displayed isn't of much help to us during analysis.

Now let's drop these three columns from our dataset (`seller`, `offer_type`, `num_photos`).

In [17]:
autos = autos.drop(["num_photos", "seller", "offer_type"], axis=1)

In [18]:
# now let's look at our new dataset
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price  ab_test vehicle_type  registration_year     gearbox  power_ps  \
0   5000  control          bus                2004    manuell       158   
1   8500  control    limousine                1997  automatik       286   
2   8990     test    limousine                2009    manuell       102   
3   4350  control   kleinwagen                2007  automatik        71   
4   1350     test        kombi                2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere       150000                   3       lpg     peugeot   
1     7er       150000                   6    benzin         bmw   
2    golf        70000                   7    benzin  volkswagen   
3  fortwo        70000                   6    benzin       smart   
4   focus       150000                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

In [19]:
autos.columns

Index(['date_crawled', 'name', 'price', 'ab_test', 'vehicle_type',
       'registration_year ', 'gearbox', 'power_ps', 'model', 'odometer_km',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'postal_code', 'last_seen'],
      dtype='object')

In [20]:
autos.describe(include="all")

date_crawled         name         price ab_test vehicle_type  \
count                 50000        50000  5.000000e+04   50000        44905   
unique                48213        38754           NaN       2            8   
top     2016-03-12 16:06:22  Ford_Fiesta           NaN    test    limousine   
freq                      3           78           NaN   25756        12859   
mean                    NaN          NaN  9.840044e+03     NaN          NaN   
std                     NaN          NaN  4.811044e+05     NaN          NaN   
min                     NaN          NaN  0.000000e+00     NaN          NaN   
25%                     NaN          NaN  1.100000e+03     NaN          NaN   
50%                     NaN          NaN  2.950000e+03     NaN          NaN   
75%                     NaN          NaN  7.200000e+03     NaN          NaN   
max                     NaN          NaN  1.000000e+08     NaN          NaN   

        registration_year   gearbox      power_ps  model    odometer_km  \
count         50000.000000    47320  50000.000000  47242   50000.000000   
unique                 NaN        2           NaN    245            NaN   
top                    NaN  manuell           NaN   golf            NaN   
freq                   NaN    36993           NaN   4024            NaN   
mean           2005.073280      NaN    116.355920    NaN  125732.700000   
std             105.712813      NaN    209.216627    NaN   40042.211706   
min            1000.000000      NaN      0.000000    NaN    5000.000000   
25%            1999.000000      NaN     70.000000    NaN  125000.000000   
50%            2003.000000      NaN    105.000000    NaN  150000.000000   
75%            2008.000000      NaN    150.000000    NaN  150000.000000   
max            9999.000000      NaN  17700.000000    NaN  150000.000000   

        registration_month fuel_type       brand unrepaired_damage  \
count         50000.000000     45518       50000             40171   
unique                 NaN         7          40                 2   
top                    NaN    benzin  volkswagen              nein   
freq                   NaN     30107       10687             35232   
mean              5.723360       NaN         NaN               NaN   
std               3.711984       NaN         NaN               NaN   
min               0.000000       NaN         NaN               NaN   
25%               3.000000       NaN         NaN               NaN   
50%               6.000000       NaN         NaN               NaN   
75%               9.000000       NaN         NaN               NaN   
max              12.000000       NaN         NaN               NaN   

                 ad_created   postal_code            last_seen  
count                 50000  50000.000000                50000  
unique                   76           NaN                39481  
top     2016-04-03 00:00:00           NaN  2016-04-07 06:17:27  
freq                   1946           NaN                    8  
mean                    NaN  50813.627300                  NaN  
std                     NaN  25779.747957                  NaN  
min                     NaN   1067.000000                  NaN  
25%                     NaN  30451.000000                  NaN  
50%                     NaN  49577.000000                  NaN  
75%                     NaN  71540.000000                  NaN  
max                     NaN  99998.000000                  NaN

In [21]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   name                50000 non-null  object
 2   price               50000 non-null  int64 
 3   ab_test             50000 non-null  object
 4   vehicle_type        44905 non-null  object
 5   registration_year   50000 non-null  int64 
 6   gearbox             47320 non-null  object
 7   power_ps            50000 non-null  int64 
 8   model               47242 non-null  object
 9   odometer_km         50000 non-null  int64 
 10  registration_month  50000 non-null  int64 
 11  fuel_type           45518 non-null  object
 12  brand               50000 non-null  object
 13  unrepaired_damage   40171 non-null  object
 14  ad_created          50000 non-null  object
 15  postal_code         50000 non-null  int64 
 16  last_seen           50

We'll keep on exploring the data as we look for data that's not alright from the dataset

We'll start by looking at our `price` and `odometer_km` columns. We will check for the following:
* Analyze the columns using minimum and maximum values and look for any values that look unrealistically ***high*** or ___low___ (outliers) that we might want to remove.
* We'll use:
    - `Series.unique().shape` - to see how many unique values
    - `Series.describe()` - to view ***min***/***max***/***median***/***mean*** etc
    - `Series.value_counts()`, with some variations:
        - chained to `.head()` if there are lots of values.
        - Because `Series.value_counts()` returns a series, we can use `Series.sort_index()` with `ascending= True or False` to view the ***highest*** and ***lowest*** values with their counts (can also chain to `head()` here).
    - When removing outliers, we can do `df[(df["col"] >= x ) & (df["col"] <= y )]`, but it's more readable to use `df[df["col"].between(x,y)]`

In [22]:
# odometer_km
# checking how many unique values are in the odometer_km series
autos["odometer_km"].unique().shape

(13,)

In [23]:
# checking for min/max/medium/mean values
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [24]:
# checking for variations
# autos["odometer_km"].value_counts().head()
# autos["odometer_km"].value_counts()
# checking countsin descending order to check for the highest and lowest values
autos["odometer_km"].value_counts().sort_index(ascending = False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

We see that the values in the `odometer_km` field are rounded, which might indicate that sellers had to choose from pre-set options for this field. We can also see that there are more high mileage than low mileage vehicles.

In [25]:
# exploring price column
autos["price"].unique().shape

(2357,)

In [26]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [27]:
autos["price"].value_counts().head()

0       1421
500      781
1500     734
2500     643
1000     639
Name: price, dtype: int64

In [28]:
autos["price"].value_counts().sort_index(ascending = True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64

In [29]:
autos["price"].value_counts().sort_index(ascending = False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

In [30]:
price_rec = autos["price"].shape[0]
print(f"Total number of 'price' records ==> {price_rec}" )
# percentage of $0 prices
zero_counts_perc = (1421 / price_rec) * 100
print()
print("$0 price percentage ==> {0}".format(zero_counts_perc))

Total number of 'price' records ==> 50000

$0 price percentage ==> 2.842


The `prices` just like the values in the `odometer_km` column seem rounded. However, it has ___2357___ unique values for the column.

As you can see, there are 1,421 cars listed with the price of $0. this only accounts for ***2.842%*** of the cars.

According to my opinion, we might consider removing these rows because how can a vehicle be sold for $0 which doesn't make sense to me.

We've also found out that the maximum price is one hundred million dollars. let's explore further and see what we can get out of the prices.

In [31]:
# checking for the high prices
autos["price"].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

In [32]:
# checking for lower prices again
autos["price"].value_counts().sort_index(ascending=True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64

Below `$35`, there are quite a number of car listings. 

eBay is normall an auction site and there could be items where the opening bid is `$1`. 

We will keep the `$1` items, but remove anything above `$350,000`, since from this increasing trend,  it seems that prices increase steadily to that number and then jump up to less realistic numbers in terms of differences in trends. Like the difference between `$350,000` and `$999,990` is `$649,990` and it keeps getting higher for the trends ans you go upwards which makes this trends look weird.

In [33]:
# removing unrealistic trends with hugher prices
autos = autos[autos["price"].between(1, 351000)]
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

In [34]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48565 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        48565 non-null  object
 1   name                48565 non-null  object
 2   price               48565 non-null  int64 
 3   ab_test             48565 non-null  object
 4   vehicle_type        43979 non-null  object
 5   registration_year   48565 non-null  int64 
 6   gearbox             46222 non-null  object
 7   power_ps            48565 non-null  int64 
 8   model               46107 non-null  object
 9   odometer_km         48565 non-null  int64 
 10  registration_month  48565 non-null  int64 
 11  fuel_type           44535 non-null  object
 12  brand               48565 non-null  object
 13  unrepaired_damage   39464 non-null  object
 14  ad_created          48565 non-null  object
 15  postal_code         48565 non-null  int64 
 16  last_seen           48

## Exploring the dates
Dtes are represented by `5` columns in our dataset, where some were created by the ___crawler___ used to mine the data and some are from the same website.

Here's the data dictionary for the dates:
- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

In our dataset,  the `date_crawled`, `last_seen` and `ad_created ` are being identified as strings by pandas. We need to covert them into numerical datain order for it to be understood quantitatively.

The other two columns that have dates are represented as numeric values which is ok for us. Let's use `Series.describe()` to be able to understand the distribution from the output received. This helps us to understand without extra data processing.

We'll use the following code syntax to explore this:

```python
autos[['date_crawled','ad_created','last_seen']][0:5]
```
The above code will give as a breif sample of the first five rows for the listed columns.

Here, you'll notice that the first 10 characters represent the day (e.g. `2016-03-12`). In order to get to know about the date range, we will extract the date values using `Series.value_counts()` to generate a distribution and then ***sort*** by ___index___.

Now, to select the first ten characters/rows/records in each column, we'll use `Series.str[:10]`:

```python
print(autos['date_crawled'].str[:10])
```

which gives:

```
0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
...
```

and so on...

In [35]:
# calculating the distribution for date_crawled, ad_created, and last_seen columns
autos[["date_crawled", "ad_created", "last_seen"]][:15]

date_crawled           ad_created            last_seen
0   2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1   2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2   2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3   2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4   2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50
5   2016-03-21 13:47:45  2016-03-21 00:00:00  2016-04-06 09:45:21
6   2016-03-20 17:55:21  2016-03-20 00:00:00  2016-03-23 02:48:59
7   2016-03-16 18:55:19  2016-03-16 00:00:00  2016-04-07 03:17:32
8   2016-03-22 16:51:34  2016-03-22 00:00:00  2016-03-26 18:18:10
9   2016-03-16 13:47:02  2016-03-16 00:00:00  2016-04-06 10:46:35
10  2016-03-15 01:41:36  2016-03-14 00:00:00  2016-04-07 03:16:21
11  2016-03-16 18:45:34  2016-03-16 00:00:00  2016-03-16 18:45:34
12  2016-03-31 19:48:22  2016-03-31 00:00:00  2016-04-06 14:17:52
13  2016-03-23 10:48:32  2016-03-23 00:00:00  2016-04-01 14:17:13
14  2016-03-23 11:50:46  2016-03-23 00:00:00  2016-04-01 15:46:47

In [36]:
(autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index())

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [37]:
(autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_values())

2016-04-07    0.001400
2016-04-06    0.003171
2016-03-18    0.012911
2016-04-05    0.013096
2016-03-06    0.014043
2016-03-13    0.015670
2016-03-05    0.025327
2016-03-24    0.029342
2016-03-16    0.029610
2016-03-27    0.031092
2016-03-25    0.031607
2016-03-17    0.031628
2016-03-31    0.031834
2016-03-10    0.032184
2016-03-26    0.032204
2016-03-23    0.032225
2016-03-11    0.032575
2016-03-22    0.032987
2016-03-09    0.033090
2016-03-08    0.033296
2016-04-01    0.033687
2016-03-30    0.033687
2016-03-29    0.034099
2016-03-15    0.034284
2016-03-19    0.034778
2016-03-28    0.034860
2016-04-02    0.035478
2016-03-07    0.036014
2016-04-04    0.036487
2016-03-14    0.036549
2016-03-12    0.036920
2016-03-21    0.037373
2016-03-20    0.037887
2016-04-03    0.038608
Name: date_crawled, dtype: float64

From the site data above, we can see that the data was being crawled on a daily basis for about one to two months between March 2016 and April 2016.

This also shows that the distribution is uniform since site was being crawled on a daily basis.

In [38]:
(autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index())

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

From the data above, we can see that most of the ads were created between 1 to 2 months across the year.

In [39]:
(autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index())

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

The site crawler recorded the date it last saw any listing, this allows us to evaluate the day in which a listing was removed, presumably because the car was sold.

We can see that the trend of distribution is consistent from the `last_seen` column which may mean that cars were being sold on a daily basis.

The last three days have an unusually high number of `last seen` values. It's unlikely that there was a massive rise in sales, and it's more likely that these values are related to the crawling period ending and do not indicate car sales.

In [40]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48565 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        48565 non-null  object
 1   name                48565 non-null  object
 2   price               48565 non-null  int64 
 3   ab_test             48565 non-null  object
 4   vehicle_type        43979 non-null  object
 5   registration_year   48565 non-null  int64 
 6   gearbox             46222 non-null  object
 7   power_ps            48565 non-null  int64 
 8   model               46107 non-null  object
 9   odometer_km         48565 non-null  int64 
 10  registration_month  48565 non-null  int64 
 11  fuel_type           44535 non-null  object
 12  brand               48565 non-null  object
 13  unrepaired_damage   39464 non-null  object
 14  ad_created          48565 non-null  object
 15  postal_code         48565 non-null  int64 
 16  last_seen           48

In [41]:
autos.columns

Index(['date_crawled', 'name', 'price', 'ab_test', 'vehicle_type',
       'registration_year ', 'gearbox', 'power_ps', 'model', 'odometer_km',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'postal_code', 'last_seen'],
      dtype='object')

In [42]:
# autos["registration_year "].describe()
# I've discovered my registration_year column name has a space after it.
# I will have to remove the space because it's bringing me an index error
autos.rename({"registration_year ": "registration_year"}, axis=1, inplace=True)
# now let's see the column names again to see If we've fixed the name problem
autos.columns

Index(['date_crawled', 'name', 'price', 'ab_test', 'vehicle_type',
       'registration_year', 'gearbox', 'power_ps', 'model', 'odometer_km',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'postal_code', 'last_seen'],
      dtype='object')

In [43]:
# it's now fixed, let's see the description now
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

From the data above, we can determine the age of a car using it. But, the data seems to be very weird. there is no way that a car was sold `1000` years ago as shown by the minimum value which is long before automobiles were invented. There is also no way that that a car was sold `9999` years into the future as indicated by the maximum value which indicates many years into the future

## Exploring incorrect registration year data
We saw that the `registration_year` column contains:

* The minimum value is `1000`, before cars were invented
* The maximum value is `9999`, many years into the future

A car cannot be registered for the first time after the listing is viewed, any vehicle with a registration year prior to `2016` is undoubtedly incorrect. Because it's more difficult to determine the earliest valid year.
Realistically, it could be in the first few decades of the twentieth century (`1900s`).

Let's count the number of listings with cars that aren't between `1900` and `2016` and see if we can safely remove those rows or if we need more custom logic.

In [44]:
# our lowest and highest possible value is 1900 and 2016 respectively
# this we'll assume in terms of percentage

(~autos["registration_year"].between(1900, 2016)).sum() / autos["registration_year"].shape[0]

0.038793369710697

This is approximately `3.8793369710697%` of our data which is less than `4%`.

We will now have to remove these rows since we won't need the for analysis.

In [45]:
# we will use the between function to avoid complicating our code
autos = autos[autos["registration_year"].between(1900, 2016)]
autos["registration_year"].value_counts(normalize=True).head(20)#.sort_index(ascending=False)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
2008    0.047450
2009    0.044665
1997    0.041794
2011    0.034768
2010    0.034040
1996    0.029412
2012    0.028063
1995    0.026285
2016    0.026135
2013    0.017202
Name: registration_year, dtype: float64

Using this, we see that most of the cars were registered in the past few years.

## Let's look at price by brand
We will use `aggregation` technique to dothe analysis of ___prices by brand___. Since we are working with cars data, it's natural to explore `variations` across different car brands.

`Aggregation` technique will help us to understand the `brand` column for this analysis.

We will use ***loops*** to to perfom `aggregation`, the process will be as follows:

* Identify the unique values we want to aggregate by
* Create an empty dictionary to store our aggregate data
* Loop over the unique values, and for each: - Subset the dataframe by the unique values - Calculate the mean of whichever column we're interested in - Assign the ***val/mean*** to the dict as ___k/v___.

In [46]:
# brands
autos["brand"].value_counts(normalize=True)#.head(15)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

Let's see which brands we are going to aggregate.

In [47]:
unique_brands = autos["brand"].unique()
unique_brands.shape[0]

40

In [48]:
(autos["brand"].value_counts(normalize=True).head().sum()) * 100

61.191919624686705

As you can see, the listing contains __German automotive companies___ in the top `5`, which contribute to over `50%` of the brand value.

`Volkswagen` is the most popular brand, with approximately double the cars for sale of the next two brands combined.

Because there are many brands with a small percentage of listings, we will limit our analysis to brands with more than `5%` of total listings. 

In [49]:
b_counts = autos["brand"].value_counts(normalize=True)
#  using Series.index attribute to access the labels
c_brands = b_counts[b_counts > 0.05].index
c_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

We chose the brands with a listing of over `5%` because they have significant percentages in the total listings.

In [50]:
brand_means = {}
for b in c_brands:
    b_name = autos[autos["brand"] == b] # fetching brand name
    b_price = b_name["price"].mean() # fetching price per brand name and adding the mean prices as they appear pwer brand
    brand_means[b] = int(b_price)# populating dict with brand_name: mean_price key:value pair
brand_means

{'volkswagen': 5402,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749}

From this dataset, this is what We've been able to gather:
* `Audi`, `BMW` and `Mercedes Benz` are more expensive compared to the others
* `Opel` and `Ford` are less expensive
* `Volkswagen` is in between - this may explain its popularity, it may be a 'best for different car categories.

Now we are going to store our aggregate data into a dataframe. We `aggregated` our data in order to understand the `mean price` for our top brands where we found a distinct price gap in the top 6 brands.

We are also going to find the `mileage` for the top 6 brands using this `aggregation` technique. This will help us to find out if there in any viable linkage to the mean price.

In [51]:
mean_mileage = {}
for m in c_brands:
    # print(m)
    m_names = autos[autos["brand"] == m]
    m_miles = m_names["odometer_km"].mean()
    mean_mileage[m] = int(m_miles)
mean_mileage

{'volkswagen': 128707,
 'bmw': 132572,
 'opel': 129310,
 'mercedes_benz': 130788,
 'audi': 129157,
 'ford': 124266}

Let's create a `series` for both dictionaries 

In [52]:
price_series = pd.Series(brand_means)
print("'Price Series'")
print(price_series)
miles_series = pd.Series(mean_mileage)
print()
print("'Mileage Series'")
print(miles_series)

'Price Series'
volkswagen       5402
bmw              8332
opel             2975
mercedes_benz    8628
audi             9336
ford             3749
dtype: int64

'Mileage Series'
volkswagen       128707
bmw              132572
opel             129310
mercedes_benz    130788
audi             129157
ford             124266
dtype: int64


Converting into a dataframe with both columns

In [53]:
brands_df_info = pd.DataFrame(price_series, columns=['mean_price'])
brands_df_info

mean_price
volkswagen           5402
bmw                  8332
opel                 2975
mercedes_benz        8628
audi                 9336
ford                 3749

In [54]:
# assigning the mileage series as a new column in the same dataframe
brands_df_info["mean_mileage"] = miles_series
brands_df_info

mean_price  mean_mileage
volkswagen           5402        128707
bmw                  8332        132572
opel                 2975        129310
mercedes_benz        8628        130788
audi                 9336        129157
ford                 3749        124266

The range of car mileages doesn't vary as much as the prices do by brand, with all major brands falling within `10%` of one another. There is a slight pattern toward higher mileage in more expensive vehicles and lower mileage in the less expensive cars.